core/trino-main/src/main/java/io/trino/cost/RemoteSourceStatsRule.java

```diff
@@ -0,0 +1,130 @@
+            PlanNodeStatsEstimate stageEstimatedStats = stageRuntimeInfoProvider.getEstimatedStats(context.statsProvider(), planFragmentId);
+            PlanNodeStatsEstimate adjustedStageStats = adjustStats(
+                    node.getOutputSymbols(),
+                    context.types(),
+                    stageRuntimeStats,
+                    stageEstimatedStats);
+
+            estimate = estimate
+                    .map(planNodeStatsEstimate -> addStatsAndMaxDistinctValues(planNodeStatsEstimate, adjustedStageStats))
+                    .or(() -> Optional.of(adjustedStageStats));
+        }
+
+        verify(estimate.isPresent());
+        return estimate;
+    }
+
+    private PlanNodeStatsEstimate adjustStats(
+            List<Symbol> outputs,
+            TypeProvider typeProvider,
+            OutputStatsEstimateResult runtimeStats,
+            PlanNodeStatsEstimate estimateStats)
+    {
+        if (runtimeStats.isUnknown()) {
+            return estimateStats;
+        }
+
+        // We prefer runtime stats over estimated stats, because runtime stats are more accurate.
+        OutputDataSizeEstimate outputDataSizeEstimate = runtimeStats.outputDataSizeEstimate();
+        PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder()
+                .setOutputRowCount(runtimeStats.outputRowCountEstimate());
+
+        double fixedWidthTypeSize = 0;
+        double variableTypeNonNullRowCount = 0;
+
+        for (Symbol outputSymbol : outputs) {
+            Type type = typeProvider.get(outputSymbol);
+            SymbolStatsEstimate symbolStatistics = estimateStats.getSymbolStatistics(outputSymbol);
+            double nullsFraction = firstNonNaN(symbolStatistics.getNullsFraction(), 0d);
+            double numberOfNonNullRows = runtimeStats.outputRowCountEstimate() * (1.0 - nullsFraction);
+
+            if (type instanceof FixedWidthType) {
+                fixedWidthTypeSize += numberOfNonNullRows * ((FixedWidthType) type).getFixedSize();
+            }
+            else {
+                variableTypeNonNullRowCount += numberOfNonNullRows;
+            }
+        }
+
+        double runtimeOutputDataSize = outputDataSizeEstimate.getTotalSizeInBytes();
+        double averageVariableSize = NaN;
+        if (variableTypeNonNullRowCount > 0 && runtimeOutputDataSize > fixedWidthTypeSize) {
+            averageVariableSize = (runtimeOutputDataSize - fixedWidthTypeSize) / variableTypeNonNullRowCount;
+        }
+
+        for (Symbol outputSymbol : outputs) {
+            SymbolStatsEstimate symbolStatistics = estimateStats.getSymbolStatistics(outputSymbol);
+            Type type = typeProvider.get(outputSymbol);
+            if (!(isNaN(averageVariableSize) || type instanceof FixedWidthType)) {
+                symbolStatistics = SymbolStatsEstimate.buildFrom(symbolStatistics)
+                        .setAverageRowSize(averageVariableSize)
```

Comment:
Isn't `averageVariableSize` average total size of variable sized symbol per-row, not average size of a single symbol?

Reply:
Actually I think this is fine - but variable is named poorly.